### PR TITLE
fix invalid timestamp flag: 4 when pushing down timestamp to coprocessor

### LINKED ISSUE #164: Invalid Flag type for TimestampType: 4
Known affected branch:
```sql
master
DAGRequestConstruct
```

SQL:
```sql
select  distinct(tp_timestamp)  from full_data_type_table  order by tp_timestamp
```

Exception:
```java
Job aborted due to stage failure: Task 0 in stage 3180.0 failed 1 times, most recent failure: Lost task 0.0 in stage 3180.0 (TID 12048, localhost, executor driver): com.pingcap.tikv.codec.InvalidCodecFormatException: Invalid Flag type for TimestampType: 4
	at com.pingcap.tikv.types.TimestampType.decodeNotNull(TimestampType.java:62)
	at com.pingcap.tikv.types.DataType.decodeValueToRow(DataType.java:133)
	at com.pingcap.tikv.row.DefaultRowReader.readRow(DefaultRowReader.java:38)
	at com.pingcap.tikv.operation.iterator.CoprocessIterator$1.next(CoprocessIterator.java:78)
	at com.pingcap.tikv.operation.iterator.CoprocessIterator$1.next(CoprocessIterator.java:74)
	at org.apache.spark.sql.tispark.TiRDD$$anon$2.next(TiRDD.scala:79)
	at org.apache.spark.sql.tispark.TiRDD$$anon$2.next(TiRDD.scala:55)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIterator.agg_doAggregateWithKeys$(Unknown Source)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIterator.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anonfun$8$$anon$1.hasNext(WholeStageCodegenExec.scala:377)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
	at org.apache.spark.shuffle.sort.BypassMergeSortShuffleWriter.write(BypassMergeSortShuffleWriter.java:126)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:96)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:53)
	at org.apache.spark.scheduler.Task.run(Task.scala:99)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:322)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

Driver stacktrace: 

```

Seems we did't handle `UINT_FLAG` type in `TimestampType`

### PR DESCRIPTION
**Pull request by birdstorm**
fix #164

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tikv-client-lib-java/181)

### CODE REVIEW

**Review by Novemser — APPROVED** [review]
LGTM

**Comment by birdstorm** [comment]
corresponding integration tests passed.

File: src/main/java/com/pingcap/tikv/types/DateTimeType.java [hunk 1]
```diff
@@ -17,18 +17,15 @@
...
 
-import java.sql.Timestamp;
-import java.time.LocalDateTime;
 import java.time.ZoneId;
-import java.time.ZonedDateTime;
 
 public class DateTimeType extends TimestampType {
-  private final ZoneId defaultZone = ZoneId.systemDefault();
+  static {
+    defaultZone = ZoneId.systemDefault();
```
Comment (ilovesoup) [inline src/main/java/com/pingcap/tikv/types/DateTimeType.java]: This share the same static ref with TimestampType. You change it one place, all affected. In other words, this won't work correctly.
You might either use a method and override it for timezone. For classname in warning message you can use this.getClass().

**Review by ilovesoup — CHANGES_REQUESTED** [review]

File: src/main/java/com/pingcap/tikv/types/DateTimeType.java (hunk 1 above)
Comment (birdstorm) [reply src/main/java/com/pingcap/tikv/types/DateTimeType.java]: you're right, I will change it.

Comment (birdstorm) [reply src/main/java/com/pingcap/tikv/types/DateTimeType.java]: fixed

File: src/main/java/com/pingcap/tikv/types/DateTimeType.java [hunk 2]
```diff
@@ -17,22 +17,21 @@
...
 
 public class DateTimeType extends TimestampType {
-  private final ZoneId defaultZone = ZoneId.systemDefault();
+  private static final ZoneId defaultZone = ZoneId.systemDefault();
   static DateTimeType of(int tp) {
     return new DateTimeType(tp);
   }
 
+  @Override
+  ZoneId getDefaultTimezone() {
```
Comment (ilovesoup) [inline src/main/java/com/pingcap/tikv/types/DateTimeType.java]: use protected

File: src/main/java/com/pingcap/tikv/types/DateTimeType.java [hunk 3]
```diff
@@ -17,22 +17,21 @@
...
 import com.pingcap.tikv.meta.TiColumnInfo;
 
-import java.sql.Timestamp;
-import java.time.LocalDateTime;
 import java.time.ZoneId;
-import java.time.ZonedDateTime;
 
 public class DateTimeType extends TimestampType {
-  private final ZoneId defaultZone = ZoneId.systemDefault();
+  private static final ZoneId defaultZone = ZoneId.systemDefault();
```
Comment (ilovesoup) [inline src/main/java/com/pingcap/tikv/types/DateTimeType.java]: To keep it consistent, we use all upper case for const. DEFAULT_ZONE = ....

File: src/main/java/com/pingcap/tikv/types/TimestampType.java [hunk 4]
```diff
@@ -29,11 +29,19 @@
 import java.util.Date;
 
 public class TimestampType extends DataType {
-  private final ZoneId defaultZone = ZoneId.of("UTC");
+  private static final ZoneId utcTimezone = ZoneId.of("UTC");
```
Comment (ilovesoup) [inline src/main/java/com/pingcap/tikv/types/TimestampType.java]: Same here.

File: src/main/java/com/pingcap/tikv/types/TimestampType.java [hunk 5]
```diff
@@ -29,11 +29,19 @@
...
   static TimestampType of(int tp) {
     return new TimestampType(tp);
   }
 
+  ZoneId getDefaultTimezone() {
```
Comment (ilovesoup) [inline src/main/java/com/pingcap/tikv/types/TimestampType.java]: same here.

**Review by ilovesoup — APPROVED** [review]
LGTM as long as comments addressed.

**Comment by birdstorm** [comment]
@ilovesoup done.

**Pull request merged**